### ISSUE TITLE
PIR Codegen Issue

### ISSUE THREAD

**Issue by blackwings-01** [stanford-ppl/spatial#49]
I will use this post to keep tack of the issue I found in apps

- [ ] For InOutArgDesign, the outer controller now is generated as UnitPipeline. But UnitPipeline is is a subclass of inner most pipeline stage. A single stage outer controller can either use sequential or metapipeline. Possible a misunderstanding here?

**Comment by blackwings-01** [comment]
- [ ] Got error 
```
uninitialized effect context: effectful statements may only be used within a reifyEffects { .. } block
```
When generation SimpleSequential

**Comment by blackwings-01** [comment]
- [ ] In dot product, write port from memory controller is `x1747_mc.vdata` instead of `x1747_mc.dataIn`
- [ ] The assumption right now is any outer controller including StreamController cannot make copy of counterChain. In dot product, StreamController x1743 made a copy

**Comment by dkoeplin** [comment]
Not able to reproduce x1747_mc.vdata bug, may be something I fixed already but wasn't seen yet due to other problems

**Comment by blackwings-01** [comment]
[PIRNetwork.pdf](https://github.com/stanford-ppl/spatial/files/591797/PIRNetwork.pdf)
The generated dot product doesn't conform with PIR ctrl assumption slightly. The Inner product pipe x1831_0 is a stage of MetaPipeline stage. The assumption is that MetaPipeline stages can only have SramOnRead and SramOnWrite. So it can contains a FIFOOnWrite with SramOnRead, which is a SemiFIFO.  Is this something that we can change in the app? For the time being, if this is hard to change in spatial I can see if I can hack something up.

**Comment by dkoeplin** [comment]
Ah right. The current version of I think TPCHQ6 and DotProduct use FIFOs instead of SRAMs to load from main memory. I'm not sure why a FIFO would need a buffering scheme, as its use still guarantees that you can overlap reading/writing at the level of single elements. 

If you need me to I can probably change the outer MetaPipeline to a stream controller, but that brings to mind another question I had. How can you tell the difference in PIR between a split Unit pipe and a split regular (parallelized) pipe? They're both just called StreamController with some StreamPipes.

**Comment by blackwings-01** [comment]
The current outer MetaPipeline in dotProduct is Fine. I'm wondering if the InnerProduct CU can be changed to reading from two SemiFIFO instead of popping from two FIFO

- [ ] Currently it's too expensive to implement divMod operation in hardware for addr calculation pipe because they usually have many stages, and divMod has to be pipelined across 4 stages. So what I'm doing now is pattern match on FixMod operation and change it to FixSra right shift if the second operand is a constant and is power of 2. Can you change the blocking size in dot product to be 64 instead? which is plasticine burst size. Currently it's 96, which is maxJ burst size.

**Comment by dkoeplin** [comment]
Yeah we can change to use SRAMs + change burst size. Will do that for all the apps and regen.

**Comment by blackwings-01** [comment]
In DotProduct, the last UnitePipeline inside sequential only does copying accumulator in MetaPipe into argOut. Raghu said this was to overcome limitation in maxJ and is explicit written in app. Is it possible to remove this for plasticine? If this can be done I will be eliminate the single stage Sequential as well
<img width="714" alt="pirnetwork_pdf__1_page_" src="https://cloud.githubusercontent.com/assets/5553709/20332551/6c7a5358-ab61-11e6-83d6-795ccefccd32.png">

**Comment by dkoeplin** [comment]
It should be yes. That just implies that you keep writing the bus with intermediate values, rather than only once with the final value. I can actually add that as an optimization rule if you'd like.

**Comment by blackwings-01** [comment]
It's okay that the accumulation register is written out multiple times. The host should not look at the value until the final token comes back. I have something that does the optimization but it's a bit nasty for me to touch all the mutable fields and rewire the counters. I've tested it for a couple of simple case. It might make sense for you to do it if you have time.

**Comment by blackwings-01** [comment]
- [ ] The newly generated dotProduct doesn't have swapWrite for the SemiFIFO. The swapWrite should just be outer counter of the streamController's counter that wraps around the MC

**Comment by blackwings-01** [comment]
- [ ] Duplicated variable name in generated MatMult_outer.scala
